### PR TITLE
feat: add switcher protocol type

### PR DESCRIPTION
**Pull request by thecode**
BREAKING CHANGE: SwitcherBridge broadcast ports changed to list

Signed-off-by: thecode <levyshay1@gmail.com>

# Pull Request

Add protocol types support to api and bridge

## Description
- Add protocol type to `DeviceType` enum
- Add `SwitcherType1Api` and `SwitcherType2Api` to api
- Fix `control_device` script to use new api
- Support bridge listening on multiple ports
- Fix `discover_devices` script to support multiple ports with user selection of ports or protocol type

## Checklist

- [x] I have followed this repository's contributing guidelines.
- [x] I have used the conventional commits specification for my commit messages.
- [x] I have signed my commits.
- [x] I will adhere to the project's code of conduct.

## Additional information

> Anything else?
>

### CODE REVIEW

File: src/aioswitcher/bridge.py
Comment (jafar-atili) [inline src/aioswitcher/bridge.py]: I thought only Breeze and Runners broadcasting on 2003

Comment (thecode) [reply src/aioswitcher/bridge.py]: Correct, this is a typo in the description, this is the real mapping for type 1 or 2:
https://github.com/TomerFi/aioswitcher/blob/8215236a5549dc44c4cbd956040c22e31911f79d/src/aioswitcher/device/__init__.py#L38-L46

File: scripts/discover_devices.py
Comment (jafar-atili) [inline scripts/discover_devices.py]: I would list here the the DeviceType.value of each type/category
"set protocol type: 
        1 [Switcher Mini, Switcher Power Plug, etc..] 
        2[Breeze, Runner, Runner Mini]

Comment (jafar-atili) [reply scripts/discover_devices.py]: and also I think --type somehow conflict with --ports
I think we should have one.

I think we should have single option to control the bridge socket:
--device-category [cat1, cat2, all]

File: scripts/discover_devices.py
Comment (jafar-atili) [inline scripts/discover_devices.py]: leftover?

**Review by jafar-atili — APPROVED** [review]

File: scripts/discover_devices.py
Comment (thecode) [reply scripts/discover_devices.py]: agree, so I will remove the option to configure the ports, I also thought at start to call it device category but it conflicts with the category we already have in the code. So maybe we keep it as `protocol type`.

I will remove the option to set the port and keep only the option for protocol type, e.g.:
`type1, type2, all`

And add text which explains which devices use each type (as you wrote above)

File: scripts/discover_devices.py
Comment (jafar-atili) [inline scripts/discover_devices.py]:
```suggestion
    Protocol type 1 (UDP port 20002), used by: {", ".join([d.value for d in DeviceType if d.protocol_type == 1])}
```

File: scripts/discover_devices.py
Comment (jafar-atili) [inline scripts/discover_devices.py]:
```suggestion
    Protocol type 2 (UDP port 20003), used by: {", ".join([d.value for d in DeviceType if d.protocol_type == 2])}
```

File: scripts/discover_devices.py
Comment (jafar-atili) [reply scripts/discover_devices.py]: We already have this data, better than write it hardcoded

File: scripts/discover_devices.py
Comment (jafar-atili) [inline scripts/discover_devices.py]:
```suggestion
_examples = f"""Executing this script will print a serialized version of the discovered Switcher
```

**Review by jafar-atili — APPROVED** [review]

File: src/aioswitcher/device/__init__.py
Comment (TomerFi) [inline src/aioswitcher/device/__init__.py:49]: what do protocol_type stands for?
can we have an enum instead?

File: src/aioswitcher/api/__init__.py
Comment (TomerFi) [inline src/aioswitcher/api/__init__.py:61]: If these are static ports, why not include them in the var name?
```python
SWITHCER_TCP_PORT_9957 = 9957
SWITCHER_TCP_PORT_10000 = 10000
```
?

File: src/aioswitcher/api/__init__.py
Comment (TomerFi) [inline src/aioswitcher/api/__init__.py:566]: Wouldn't it be better to have just one API class that takes the TCP port in its constructor?
Even and enum would be good I believe as a port identifier.

And if not, can we think of a title to better categorize _plugs & boilers_ vs. _thrmostats & runners_ instead of _type1_ vs. _type2_?

This just sounds too weird, if it's mandatory maybe consider boiler-plating this by creating similar classes to identify various types?

File: src/aioswitcher/api/packets.py
Comment (TomerFi) [inline src/aioswitcher/api/packets.py:23]: This still bothers me, we talked about it in the previous PRs, yet we merged it as-is.
Can we do better the TYPE1 vs TYPE2 for all the constants in this file?

If not, maybe consider a dynamic constants script injected by the initiator...

File: src/aioswitcher/bridge.py
Comment (TomerFi) [inline src/aioswitcher/bridge.py:281]: going back to my common theme, can we do better than type1 vs type2?
we can even inject the get IP function based on the device type to make it more clear, no?

**Review by TomerFi — CHANGES_REQUESTED** [review]

File: scripts/discover_devices.py
Comment (TomerFi) [inline scripts/discover_devices.py:90]: the type1/2 is frustrating...
a better categorizing title would be much better.

**Review by TomerFi — CHANGES_REQUESTED** [review]

File: tests/test_bridge.py
Comment (TomerFi) [inline tests/test_bridge.py:53]: Shouldn't we **also** use multiple ports to make this and other tests in this file better reflect the modification in this PR?
I think we should add tests that sort, or perhaps a better idea will be to parameterize the current test cases...

**Review by TomerFi — CHANGES_REQUESTED** [review]

File: tests/test_bridge.py
Comment (thecode) [reply tests/test_bridge.py:53]: I agree, tests should be added, this why I was aiming that we work on a branch so we can split this into multiple PRs, so a PR that add tests coverage should be made before we make the final merge to main.

File: src/aioswitcher/device/__init__.py
Comment (thecode) [reply src/aioswitcher/device/__init__.py:49]: Yes, can you suggest something that will reflect the two protocol types? also note that we later had classes that are named Type1 and Type2

File: src/aioswitcher/api/__init__.py
Comment (thecode) [reply src/aioswitcher/api/__init__.py:61]: Because the Type1 and Type2 aligns with the two protocol types in all other places. When you name it `SWITCHER_TCP_PORT_10000` it does not explain which devices or protocols related to this port. it only explains that it uses port 10000 which is irrelevant because if it was the only usage we could just eliminate this constant and set the port directly.

File: src/aioswitcher/api/packets.py
Comment (thecode) [reply src/aioswitcher/api/packets.py:23]: I am sorry I am not following you, we had a discussion that we agreed that we split all places that have to TYPE1 & TYPE2 so that it is clear that there are two types of protocols. As far as I see it if you don't agree with this base assumption (I see it repeating in comments). We should close this PR, discuss again and make a new one.

File: src/aioswitcher/bridge.py
Comment (thecode) [reply src/aioswitcher/bridge.py:281]: Same as above....

File: scripts/discover_devices.py
Comment (thecode) [reply scripts/discover_devices.py:90]: Same...

File: src/aioswitcher/api/__init__.py
Comment (TomerFi) [reply src/aioswitcher/api/__init__.py:61]: Yeah but the port itself is just an integer and is not directly related to the type of device using it as its broadcast port.
For future contributors, or even for our future self, "SWITCHER_TCP_PORT2" doesn't have too much meaning to it if you're not aware and/or make the connection that this correlated with the TYPE2 device.

File: src/aioswitcher/device/__init__.py
Comment (TomerFi) [reply src/aioswitcher/device/__init__.py:49]: No, I actually can't think of anything...

File: src/aioswitcher/api/packets.py
Comment (TomerFi) [reply src/aioswitcher/api/packets.py:23]: I remember discussing it, I remember going along with this.
But I also remember constantly raising my concerns in regards to the naming issues with the new contributions,
and I remember getting constantly reassured that this is a WIP branch and we can address my naming concerns later.

I'll reiterate my concerns:
to anyone who's not involved with this upcoming release,
stuff like TYPE1/2, FORMAT1/2, and PORT1/2 don't mean much.

I understand the TYPE types and variables are a bit tricky to handle, AFAIK every type represents a set of devices that seemingly doesn't have a specific shared property that causes them to be in a specific TYPE group.
But I think that at the very least the PORT and PACKET_FORMAT variables can be named better.
Although, I would prefer the TYPEs to have even more meaning as well.

File: tests/test_bridge.py
Comment (TomerFi) [reply tests/test_bridge.py:53]: Working on a WIP branch and "adding tests later" is ok, because it's relatively easy to locate missing tests.
However, realizing you need to modify an existing working test by just adding another test case to make it cover another end case,
I believe this is more easily noted before merging to the WIP branch unless, of course, you already took note of this and plan to include it when adding the new tests.
:-)

File: src/aioswitcher/api/packets.py
Comment (thecode) [reply src/aioswitcher/api/packets.py:23]: Can you suggest something with better meaning? The easiest way would be to review the changes and add code suggestions so I can commit them if they look good.
I don't have any better approach I can think of than what implemented here.

File: tests/test_bridge.py
Comment (thecode) [reply tests/test_bridge.py:53]: I don't need to modify tests, the failing test is due to code coverage which is already missing on the base branch not on this PR. I can't do anything here until the tests PR is merged and rebasing this PR. This PR does not have any effect on coverage.
image

Comment (TomerFi) [reply tests/test_bridge.py:53]: Note my initial https://github.com/TomerFi/aioswitcher/pull/484#discussion_r897784440 :
> Shouldn't we **also** use multiple ports

meaning, adding another test case regardless of failing/passing tests.

File: src/aioswitcher/api/packets.py
Comment (TomerFi) [reply src/aioswitcher/api/packets.py:23]: > Can you suggest something with better meaning? The easiest way would be to review the changes and add code suggestions so I can commit them if they look good.

So... you don't take reviews without alternative suggestions...
For me to be able to intervene further on a expected level, besides the suggestion I had already made in regards to the port variables naming, which you opposed, means I would have to get more invested in the next release, much like you and Jafar already are, and unfortunaly, I don't think I can spare the time in foreseen future.

You could have specified why is this such an issue and perhaps we could have brainstormed together,
or you could have stated that you've considered everything else but that's the only way and you own it (you are a maintainer here no?), instead, you took the if-you-don't-like-it-then-rewrite-it approach.

Dude, as I told you previously...
you are a maintainer here,
you maintain the main client of this repo (ha),
you are more involved with the users than I am and therefore know what is requested,
you are (partially) connected to the vendor,
you were involved in the planning of these modifications with Jafar.
I don't have the time to catch up on everything **you already know** in regards to this project,
so from my end, **whatever you say goes** - you have my blessing and I would try to help out whichever way I can.

To this, you replied that I can help out with code reviews,
yet it looks like you don't really like my code reviews.
:man_shrugging:

So...
Let's just keep this as-is,
I'll approve and merge this now.

**Pull request merged**
